woocommerce/woocommerce
# Replace the Cost of Goods Sold "overrides parent" flag for variations with an "is additive" flag

### Changes proposed in this Pull Request:

This is a modification to the handling of Cost of Goods Sold values for variations that was introduced in https://github.com/woocommerce/woocommerce/pull/51675.

Currently, the cost value for variations is additive (the effective value is the sum of the parent product value and the value defined for the variation), but there's a per variation "cost overrides parent" flag that makes the defined value to become the final value for the variation.

This pull request changes the name and reverses the meaning of the flag:

* Its name is now "is additive"
* When false (default) the defined value for the variation is its final cost value; when true, the final cost value is the sum of the parent value and the defined value for the variation.

This translates in changes in the codebase as follows:

* The methods to handle the flag in `WC_Product_Variation` are now named `get_cogs_value_is_additive` and `set_cogs_value_is_additive`.
* The name of the meta key where the flag is stored is now `_cogs_value_is_additive`.
* The name of the filters that fire when the flag is loaded or saved are now `woocommerce_load_product_cogs_is_additive_flag` and `woocommerce_save_product_cogs_is_additive_flag`.
* The name of the field in the JSON data for REST API requests is now `defined_value_is_additive`.
* And of course, the logic in `WC_Product_Variation::get_cogs_effective_value_core` is the reverse as with the old "overrides" flag.

The motivation for this change is some research that seems to suggest that variations having their own absolute cost values will be the most common use case.


### How to test the changes in this Pull Request:

Same testing instructions of https://github.com/woocommerce/woocommerce/pull/51675 with the corresponding changes:

* For step 4, the expected values change:

```php
$v=wc_get_product($variation_id);

$v->set_cogs_value(20.5);

echo $v->get_cogs_value(); // 20.5
echo $v->get_cogs_effective_value(); // 20.5
echo $v->get_cogs_total_value(); // 20.5
```

* Same for steps 5 and 6, also note the changes in the names of methods and keys:

```php
echo $v->get_cogs_value_is_additive(); // false
$v->set_cogs_value_is_additive(true);
echo $v->get_cogs_value_is_additive(); // true

echo $v->get_cogs_value(); // 20.5
echo $v->get_cogs_effective_value(); // 32.84
echo $v->get_cogs_total_value(); // 32.84

$v->save();

echo get_post_meta($variation_id, '_cogs_value_is_additive', true); // 'yes'

$v->set_cogs_value_is_additive(false);
$v->save();

global $wpdb; echo $wpdb->get_var("SELECT COUNT(1) FROM {$wpdb->postmeta} WHERE meta_key='_cogs_value_is_additive' AND post_id=$variation_id"); // 0
```

* For the REST API tests, use `"defined_value_is_additive"` as the field name for the flag and verify that the effective cost values returned are consistent with the new semantics; verify also the description returned by an `OPTIONS` query for the field corresponding to the flag.

* For the hooks tests, use the names `woocommerce_load_product_cogs_is_additive_flag` and `woocommerce_save_product_cogs_is_additive_flag` instead of the names with `...overrides_parent_value...`.


### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
